### PR TITLE
Add function attributes to functions in dmd/root

### PR DESCRIPTION
**Pull request by dkorpel**
Point 1 in [DMD best practices](https://github.com/dlang/dmd/blob/master/CONTRIBUTING.md#dmd-best-practices). Found using [dlp](https://github.com/jacob-carlborg/dlp).

### CODE REVIEW

File: src/dmd/root/bitarray.d [hunk 1]
```diff
@@ -44,23 +44,23 @@ nothrow:
...
-    void opAssign(const ref BitArray b)
+    void opAssign(const ref BitArray b) pure
     {
         if (!len)
             length(b.len);
         assert(len == b.len);
         memcpy(ptr, b.ptr, bytes(len));
     }
 
-    bool opIndex(size_t idx) const pure nothrow @nogc
```
Comment (dkorpel) [inline src/dmd/root/bitarray.d:55]: This `nothrow` is redundant with the `nothrow:` at the top.

File: src/dmd/root/bitarray.d [hunk 2]
```diff
@@ -44,23 +44,23 @@ nothrow:
...
+    bool opIndex(size_t idx) const pure @nogc
     {
         import core.bitop : bt;
 
         assert(idx < len);
         return !!bt(ptr, idx);
     }
 
-    void opIndexAssign(bool val, size_t idx) pure nothrow @nogc
+    void opIndexAssign(bool val, size_t idx) pure @nogc
```
Comment (dkorpel) [inline src/dmd/root/bitarray.d:63]: Same

File: src/dmd/root/bitarray.d [hunk 3]
```diff
@@ -44,23 +44,23 @@ nothrow:
         len = nlen;
     }
 
-    void opAssign(const ref BitArray b)
+    void opAssign(const ref BitArray b) pure
```
Comment (RazvanN7) [inline src/dmd/root/bitarray.d:47]: This looks like it could be `@nogc` as well, if `length` is `@nogc`.

File: src/dmd/root/bitarray.d [hunk 4]
```diff
@@ -118,7 +118,7 @@ nothrow:
         }
     }
 
-    ~this() pure nothrow
+    ~this() pure
```
Comment (RazvanN7) [inline src/dmd/root/bitarray.d:121]: Could this also be made `@nogc` ?

**Review by RazvanN7 — COMMENTED** [review]
A few comments on BitArray:

 - it seems like `pure` could also be use en-masse, like `nothrow`
 - if `opAssign` is marked `@nogc` then we could use it en-masse
 - you should also try to annotate the unittest that is present in the file

**Comment by dkorpel** [comment]
>     * it seems like `pure` could also be use en-masse, like `nothrow`

I saw that, but the style guide prefers individual attributes on each function.

>     * if `opAssign` is marked `@nogc` then we could use it en-masse

At first I marked `@nogc` where I could, but the auto-tester failed on configurations with the GC enabled where `mem.free` and co aren't `@nogc`, so I removed them.

>     * you should also try to annotate the unittest that is present in the file

I'll see

**Comment by RazvanN7** [comment]
> I saw that, but the style guide prefers individual attributes on each function.

Well, we should stick with one or the other, not have both.

**Comment by dkorpel** [comment]
> Well, we should stick with one or the other, not have both.

So should I remove `nothrow:` or add `pure:`?

**Comment by RazvanN7** [comment]
It seems that either way we end up with a significant amount of changes. I, personally, do not like the `attribute:` syntax as it is complicated to review code that contains such artifacts. Since the style checkers prefers individual attributes, I suggest adding them individually, however, I am flexible in this regard.

**Comment by RazvanN7** [comment]
Most of the occurences of `nothrow:` are coming from the backend. That makes sense, it could not possibly throw, since it was translated from C, so I would say that your argument is not convincing.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**